mlos_bench/mlos_bench/services/remote/azure/azure_auth.py

```diff
@@ -53,18 +56,60 @@ def __init__(self,
         self._access_token = "RENEW *NOW*"
         self._token_expiration_ts = datetime.datetime.now()  # Typically, some future timestamp.
 
+        # Login as ourselves
+        self._cred = azure_id.AzureCliCredential()
+
+    def _init_sp(self):
+        # Perform this initialization outside of __init__ so that environment loading tests
+        # dont need to specifically mock keyvault interactions out
+
+        # Already logged in as SP
+        if isinstance(self._cred, azure_id.CertificateCredential):
+            return
+
+        check_required_params(
+            self.config, {
+                "spClientId",
+                "keyVaultName",
+                "certName",
+                "tenant",
+            }
+        )
+
+        sp_client_id = self.config["spClientId"]
+        keyvault_name = self.config["keyVaultName"]
+        cert_name = self.config["certName"]
+        tenant_id = self.config["tenant"]
+
+        # Get a client for fetching cert info
+        keyvault_secrets_client = SecretClient(
+            vault_url=f"https://{keyvault_name}.vault.azure.net",
+            credential=self._cred,
+        )
+
+        # The certificate private key data is stored as hidden "Secret" (not Key strangely)
+        #  in PKCS12 format, but we need to decode it.
+        secret = keyvault_secrets_client.get_secret(cert_name)
+        assert secret.value is not None
+        cert_bytes = b64decode(secret.value)
+
+        # Reauthenticate as the service principal.
+        self._cred = azure_id.CertificateCredential(tenant_id=tenant_id, client_id=sp_client_id, certificate_data=cert_bytes)
+
     def get_access_token(self) -> str:
         """
         Get the access token from Azure CLI, if expired.
         """
+        # Ensure we are logged as the Service Principal, if provided
+        if "spClientId" in self.config:
+            self._init_sp()
+
         ts_diff = (self._token_expiration_ts - datetime.datetime.now()).total_seconds()
         _LOG.debug("Time to renew the token: %.2f sec.", ts_diff)
         if ts_diff < self._req_interval:
             _LOG.debug("Request new accessToken")
-            # TODO: Use azure-identity SDK and a key valut instead of `az` CLI.
-            res = json.loads(subprocess.check_output(
-                'az account get-access-token', shell=True, text=True))
-            self._token_expiration_ts = datetime.datetime.fromisoformat(res["expiresOn"])
-            self._access_token = res["accessToken"]
+            res = self._cred.get_token("https://management.azure.com/.default")
+            self._token_expiration_ts = datetime.datetime.fromtimestamp(res.expires_on)
```

Comment:
Please validate that this is in UTC or doesn't have other sorts of timezone issues.

Reply:
Good point. We get the expiration time as a [Unix timestamp](https://learn.microsoft.com/en-us/python/api/azure-core/azure.core.credentials.accesstoken?view=azure-python#azure-core-credentials-accesstoken-expires-on) but the usual datetime operations are in local timezone. Changed the datetime operations to their corresponding UTC versions!